lib/collection/proxy-config.js

```diff
@@ -0,0 +1,230 @@
+             */
+            match: MATCH_ALL_URLS,
+
+            /**
+             * The server url or the proxy url.
+             * @type {Url}
+             */
+            server: new Url(),
+
+            /**
+             * This represents whether the tunneling needs to done while proxying this request.
+             * @type Boolean
+             */
+            tunnel: false
+        });
+
+        this.update(options);
+    }), Property);
+
+_.assign(ProxyConfig.prototype, /** @lends ProxyConfig.prototype */ {
+    /**
+     * Updates the properties of the proxy object based on the options provided.
+     *
+     * @param {ProxyConfig~definition} options The proxy object structure.
+     */
+    update: function (options) {
+        if (!_.isObject(options)) {
+            return;
+        }
+        _.has(options, 'server') && this.server.update(options.server);
+        _.isString(options.match) && (this.match = options.match);
+        _.has(options, 'server') && (this.tunnel = _.isBoolean(options.tunnel) ? options.tunnel : false);
+    },
+
+    /**
+     * Tests the url string with the match provided.
+     * Follows the https://developer.chrome.com/extensions/match_patterns pattern for pattern validation and matching
+     *
+     * @param {String=} [urlStr] The url string for which the proxy match needs to be done.
+     */
+
+    test: function(urlStr) {
+        /*
+        * This function executes the code in the following sequence for early return avoiding the costly regex matches.
+        * To avoid most of the memory consuming code.
+        * 1. It check whether the match string is <all_urls> in that case, it return immediately without any further
+        *    processing.
+        * 2. Checks whether the matchPattern follows the rules, https://developer.chrome.com/extensions/match_patterns,
+        *    If not then, dont process it. It is a Regex match process which is slower, @todo needs to be cached in
+        *    future.
+        * 3. Check for the protocol, as it is a normal array check.
+        * 4. Checks the host, as it doesn't involve only string comparisons.
+        * 5. Finally, checks for the path, which actually involves the Regex matching, the slow process.
+        */
+        var matchRegexObject = {},
+            match = null,
+            url,
+            hostSuffix = '';
+
+        // If the matchPattern is <all_urls> then there is no need for any validations.
+        if (this.match === MATCH_ALL_URLS) {
+            return true;
+        }
+
+        // Check the match pattern of sanity and split it into protocol, host and path
+        match = this.match.match(regexes.patternSplit);
+
+        if (!match) {
+            // This ensures it is a invalid match pattern
+            return false;
+        }
+
+        // If the protocol retuns '*' we are going to assign ['https', 'http'] array to it.
+        matchRegexObject.protocol = match[1] === MATCH_ALL ? ALLOWED_PROTOCOLS : [match[1]];
```

Comment:
you can make a new function that will directly return  a `matchRegexObject`, so that you can test it out separately in the unit tests.